feat(lang): add R

Ben10164 opened the pull request:
I spent the last few days making a seamless LazyVim compatible implementation of [R.nvim](https://github.com/R-nvim/R.nvim). This was something requested in by #3126.

I have also never contributed to this repo, and only recently started using neovim, so please let me know if there is anything I can do!

<details>

<summary>Features</summary>

As `R.nvim` is a feature rich plugin, I figured it would be nice to demonstrate its integration with LazyVim.

## Completion
<img width="388" alt="Screenshot 2024-05-21 at 5 21 08 PM" src="https://github.com/LazyVim/LazyVim/assets/54869170/46789d95-223f-40f2-9c5b-ec95cf4e5915">

## Linting
<img width="492" alt="Screenshot 2024-05-21 at 5 24 27 PM" src="https://github.com/LazyVim/LazyVim/assets/54869170/d605e70e-a053-423d-ac6e-47eb746d6963">

## REPL integration
<img width="823" alt="Screenshot 2024-05-21 at 5 20 36 PM" src="https://github.com/LazyVim/LazyVim/assets/54869170/2136b50a-f04e-4bd9-be5c-b50c0584b0c4">

## Mason LSP
<img width="696" alt="Screenshot 2024-05-21 at 5 15 35 PM" src="https://github.com/LazyVim/LazyVim/assets/54869170/2a05e74c-3f25-467d-a4fc-95db0f07fe09">

## R Markdown
Integrates with `extras.lang.markdown`
<img width="825" alt="Screenshot 2024-05-21 at 5 22 21 PM" src="https://github.com/LazyVim/LazyVim/assets/54869170/ddbd9e20-c4d9-47be-9d45-a6b393403e93">

## Knitting support
<img width="740" alt="Screenshot 2024-05-21 at 5 22 39 PM" src="https://github.com/LazyVim/LazyVim/assets/54869170/7607d8e8-9d49-4233-b9ae-fd5f4f1e450a">

## WhichKey support
<img width="825" alt="Screenshot 2024-05-21 at 5 21 42 PM" src="https://github.com/LazyVim/LazyVim/assets/54869170/e690039d-070e-4dd6-aa3d-e128cedc473e">
</details>

Aman9das commented:
doesn't this override a lot of keybinds?

dpetka2001 commented:
I don't do any `R`, so take everything with a grain of salt. I saw that there is already #1232 and that one also includes an implementation for `neotest`. Maybe you'd like to include that as well if it's compatible with your PR? Also, just a suggestion, but you could entirely remove the `mason.nvim` spec, since the LSP servers get automatically installed if they are defined in `nvim-lspconfig` (via `mason-lspconfig`). I'm talking about these lines
```lua
{
    "williamboman/mason.nvim",
    opts = function(_, opts)
      vim.list_extend(opts.ensure_installed, {
        "r-languageserver",
      })
    end,
  },
```
These could be removed.

Ben10164 commented:
>  I saw that there is already #1232 and that one also includes an implementation for `neotest`.

I'll experiment with implementing their config into mine. I'll update the PR when I'm done, but it shouldn't take too long to, at the least, implement neotest through the `testthat` r package.

Ben10164 commented:
@dpetka2001 
I tried using the tests that was implemented in #1232, and I was able to get it to work with the official repo for `testthat`, but doesn't detect in-line tests. Im not very familiar with the project structure of testthat, but the Neovim plugin only supports the official `tests/testthat/test-*.R` syntax. It's up to you if it should be added since it does seem to work, but lacks in-line tests, something that R users are a big familiar with from R-studio.

<img width="362" alt="Screenshot 2024-05-22 at 7 00 47 PM" src="https://github.com/LazyVim/LazyVim/assets/54869170/cccdffe9-b1ec-4b22-9cf4-8c84e33a1fc1">

Aman9das commented:
I think it'd better to not hardcode colors for lualine, rather use other indicators like *, bold, italic, underline.

Ben10164 commented:
> I think it'd better to not hardcode colors for lualine, rather use other indicators like *, bold, italic, underline.
> 
> Also is overriding the lualine config in the scope of this plugin.. extending just `lualine_y` would be better in my opinion.

Im sorry, I'm a bit new to Neovim and its configs, and Lua in general. Do you think you could give an example of how I would extend lualine? Of course you don't have to do it all, but a basic example would be really nice! I'd love to get your changes implemented ASAP

Aman9das commented:
I'm new too, but this is an example:

```lua
{
  "nvimtools/none-ls.nvim",
  optional = true,
  opts = function(_, opts)
    local nls = require("null-ls")
    opts.sources = vim.list_extend(opts.sources or {}, {
      nls.builtins.diagnostics.markdownlint,
    })
  end,
}
```

Ben10164 commented:
@Aman9das Perfect! Thank you so much! I'll get working on that right now, doing the same for other plugins that also can be extended rather than instantiated.

Ben10164 commented:
I added italics for whenever something was loading or starting, and bold for when something is waiting for input/done. I also removed the colors, but to be honest, I kind of liked the colors since it gave more context as to what it signifies. The colors and status are based on a dot in the bottom left of an R project window in RStudio, so users are only familiar with a color, not necessarily any sort of character.

Attached is a video demonstrating both of them.

https://github.com/LazyVim/LazyVim/assets/54869170/947a5ddc-73af-4e77-a58b-1a1d84b09455

If there is a specific standard that LazyVim uses for the formatting of different status indicators, lmk.

Aman9das commented:
Agreed, should keep as generic as possible, while enabling as many features..

I think you should justify these changes too

```
min_editor_width = 72,
        rconsole_width = 78,
        disable_cmds = {
          "RClearConsole",
          "RCustomStart",
          "RSPlot",
          "RSaveClose",
        },
        synctex = false,
        pdfviewer = "open",
```

pdf viewer should be xdg-open, right?

dpetka2001 commented:
@Ben10164 

First of all I'm just a regular user like you, so everything I say should be just taken as my personal opinion and suggestion. It's up to the maintainer, who has the final say.

I'm not familiar with `R` at all, so about `neotest` it's up to you. Choose whatever you think is more spread out among `R` users.

Another suggestion from my part would be to incorporate what you have in the `config` function into the `opts` of the plugin spec. This makes it easier for users to change default values without having to overwrite the whole `config` function in their personal configuration.

Ben10164 commented:
@Aman9das 
> I think you should justify these changes too

Removed them in latest commit. That was part of my personal config.

> pdf viewer should be xdg-open, right?

Changed to be is dependent in latest commit. Now Windows and Mac (Debian) use `open` and linux uses `xdg-open`. 

---
@dpetka2001 
> I'm not familiar with `R` at all, so about `neotest` it's up to you. Choose whatever you think is more spread out among `R` community.

`testthat` is the most popular unit testing package for R. It's very similar to pytest in the way that it can be used in-line, but also in a separate, more isolated `tests/` directory. I think it is fine to just leave the `neotest-testthat` in there. Although in-line testing is more common, having the option to support directory structured testing (which is technically how testthat is intended to be used) doesn't hurt anybody.

> Another suggestion from my part would be to incorporate what you have in the `config` function into the `opts` of the plugin spec for the `R.nvim` plugin.

Fixed in latest commit

dpetka2001 commented on lua/lazyvim/plugins/extras/lang/r.lua:
```diff
@@ -7,64 +7,53 @@ return {
+        end,
+      },
+      pdfviewer = (function()
+        if vim.loop.os_uname().sysname:find("Windows", 1, true) then
+          return "open" -- according to @Aman9das on Github
+        elseif vim.loop.os_uname().sysname:find("Darwin", 1, true) then
+          return "open"
+        else
+          return "xdg-open" -- according to @Aman9das on Github
+        end
+      end)(),
```
Instead of all these, you could use `vim.ui.open` I believe. It returns the default open handler based on the OS. See `:h vim.ui.open()` for more info.

Ben10164 replied:
That's a great idea, but sadly the `pdfviewer` parameter only takes a path to an application (or something defined in `$PATH`). Since `vim.ui.open()` is new in vim 0.10, I wonder if there are any plans on incorporating it into `R.nvim` directly.

I also checked out `vim.ui.open()` and didn't see any direct way to get the application it uses to open, which would be really nice for this scenario.

<img width="609" alt="Screenshot 2024-05-23 at 4 11 14 PM" src="https://github.com/LazyVim/LazyVim/assets/54869170/36d6b0bd-0132-49c6-954d-e44fabe3407b">

dpetka2001 replied:
Not entirely sure myself to be honest, but could you try
```lua
      pdfviewer = (function()
        return vim.ui.open
      end)(),
```
Otherwise, maybe the maintainer will have a better idea if this is possible.

Ben10164 replied:
That was a great idea, but sadly not 😞 

<img width="605" alt="Screenshot 2024-05-24 at 11 39 25 PM" src="https://github.com/LazyVim/LazyVim/assets/54869170/13f131b8-4c64-4135-b136-783875d8f183">

dpetka2001 replied:
I don't know then, sorry. Waiting for maintainer's input would the best :)

dpetka2001 replied:
Could you also try `pdfviewer = vim.ui.open` or `pdfviewer = vim.ui.open()`? Just throwing out ideas :stuck_out_tongue_closed_eyes:

Ben10164 replied:
Should we ship it as is with the longer, manually specified per-OS code, or hopefully wait for their implementation? I've made a [PR](https://github.com/R-nvim/R.nvim/pull/144) for `R.nvim` to bring up this idea with a simple implementation that hopefully I can work with some people to properly implement it.

I am fine shipping it now and then creating a new PR to update LazyVim then when/if I implement `vim.ui.open` in the `R.nvim`.

dpetka2001 replied:
You just have to wait for the maintainer's approval. I'm just a regular user like you.

Aman9das replied:
I think shipping this asis is good. Then working on issues filed against the extra will be more effective way of improving it.

folke commented:
I've refactored the code a bit.

Would be great if you could verify/test so I can merge.

folke commented on lua/lazyvim/plugins/extras/lang/r.lua:
```diff
@@ -44,20 +39,15 @@ return {
   },
   {
     "hrsh7th/nvim-cmp",
-    dependencies = { "R-nvim/cmp-r" },
```
Ouch, that's a mistake indeed. Not home right now

folke replied:
Added it back.

Aman9das commented:
Works as intended in quarto. Very good integration. The pdfviewer issue is there if zathura isn't preinstalled on linux, but its better handled upstream

Ben10164 commented:
My PR for setting `pdfviewer = ""` to use `vim.ui.open` is being finalized.

Also a question I have for LazyVim PRs like this. Is there anything else I need to do after this? Should I make a PR to the docs repo, or is that automatically generated?

folke commented:
I only just noticed you're the same person :)

Yes, docs are auto-generated.

Let me know when it's ready to merge

Ben10164 commented:
Hey @folke ! I saw you in the other PR too!

That PR was merged a few hours ago. Should I make a change to this PR to set the value of `pdfviewer` to be blank so it opens with vim.ui.open? R.nvim still defaults to applications that might not be installed that support SynTeX and throws errors if they aren't there, so I was thinking it would be good to have a configuration that works out of the box no matter what, that can easily be configured to use SynTeX or any other application.

folke commented:
Yes, better to set it to `""` I guess

Aman9das commented:
which-key cuts off a few descriptions:

folke commented:
@Aman9das I'm fine with adding that quarto group, but overriding which-key settings to not cut-off the descriptions, is something the user needs to configure themselves. The extra should not change those which-key settings

Ben10164 commented:
@folke I believe it is ready to merge!
